### PR TITLE
OpenStack Sample App

### PR DESCRIPTION
**Pull request by Caleb-Hurshman**
Adds a Terraform script to launch a GCP instance with OpenStack (devstack), the `openstack-exporter`, and the Grafana agent installed.

#### Note: The start-up script for the vm will take ~30 minutes to run

### CODE REVIEW

File: openstack/terraform.tfstate.backup [hunk 1]
```diff
@@ -0,0 +1,292 @@
+{
```
Comment (BominRahmani) [inline openstack/terraform.tfstate.backup:1]: You don't want this file commited.

File: openstack/terraform.tfstate [hunk 2]
```diff
@@ -0,0 +1,292 @@
+{
```
Comment (BominRahmani) [inline openstack/terraform.tfstate:1]: You do not want this file commited.

**Review by BominRahmani — CHANGES_REQUESTED** [review]
Couple of things, Also can you run a `terraform fmt`. Im not sure if it matters, but i also notice a lot of the files do not have an extra line at the end.

File: openstack/variables.tfvars [hunk 3]
```diff
@@ -0,0 +1,10 @@
+gcp_auth_file="<Path/to/your/gcp/service-account/credentials>"
```
Comment (BominRahmani) [inline openstack/variables.tfvars:1]: Is there a reason this one doesn't highlight the `gcp_auth_file` in blue? It looks like the variable declaration of all variables following are highlighted in blue.

**Review by BominRahmani — APPROVED** [review]
LGTM!

**Comment by Caleb-Hurshman** [comment]
@Dasomeone On the subject of Multipass, I was initially attempting to build this sample to run with Multipass but ran into issues with executing on an M2 Mac. Essentially the `nova` service that gets installed by devstack is somewhat supported, but not officially supported, making setup on a Multipass VM impossible for me to test on my machine.

**Comment by Dasomeone** [comment]
> @Dasomeone On the subject of Multipass, I was initially attempting to build this sample to run with Multipass but ran into issues with executing on an M2 Mac. Essentially the `nova` service that gets installed by devstack is somewhat supported, but not officially supported, making setup on a Multipass VM impossible for me to test on my machine.

That's a very fair point then! Might be worth adding a small known issues/platform support/etc section to the readme just to explain that, in case we attempt to convert it in the future

File: openstack/loadgen.tf [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+      openstack router delete test-router-scale-$i
+    done
+    echo "Scaling complete"
+  }
+
+  # Call functions
+  scale_resources
+  sleep $SLEEP_DURATION
+  delete_resources
+  sleep $SLEEP_DURATION
```
[the remark this answers is not in the document]
Comment (Caleb-Hurshman) [reply openstack/loadgen.tf:91]: @Dasomeone This aspect of the loadgen script was to generate variance in the dashboard panels. Without these two functions, all the timeseries would be flat lines, so this pattern of creating and deleting resources creates a more visually interesting result when the user looks at the mixin. If that's not a priority then I'm happy to modify the script.

[the remark this answers is not in the document]
Comment (Dasomeone) [reply openstack/loadgen.tf:104]: Hmm that does make sense, but maybe we could reduce it to a shorter interval like 2-2.5 minutes? Either way I'm nitpicking here and it's good enough as-is.

File: openstack/main.tf [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    "prometheus_username" = var.prometheus_username
+    "prometheus_password" = var.prometheus_password
+    "prometheus_url"      = var.prometheus_url
+  })
+  network_interface {
+    network    = google_compute_network.vpc.name
+    subnetwork = google_compute_subnetwork.network_subnet.name
+    access_config {}
+  }
+}
```
Comment (Dasomeone) [inline openstack/main.tf:29]: Newline

File: openstack/ubuntu-versions.tf [hunk 6]
```diff
@@ -0,0 +1,5 @@
+variable "ubuntu_image" {
+  type        = string
+  description = "SKU for Ubuntu 22.04 LTS"
+  default     = "projects/ubuntu-os-cloud/global/images/ubuntu-2204-jammy-v20240208"
+}
```
Comment (Dasomeone) [inline openstack/ubuntu-versions.tf:5]: newline

File: openstack/provider-variables.tf [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+variable "gcp_project" {
+  type        = string
+  description = "GCP project name"
+}
+# define GCP zone name
+variable "gcp_zone" {
+  type        = string
+  default     = "us-central1-a"
+  description = "GCP zone name"
+}
```
Comment (Dasomeone) [inline openstack/provider-variables.tf:21]: newline

File: openstack/provider-main.tf [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+    }
+  }
+}
+# Define GCP provider
+provider "google" {
+  credentials = file(var.gcp_auth_file)
+  project     = var.gcp_project
+  region      = var.gcp_region
+  zone        = var.gcp_zone
+}
```
Comment (Dasomeone) [inline openstack/provider-main.tf:16]: newline

File: openstack/network-variables.tf [hunk 9]
```diff
@@ -0,0 +1,5 @@
+variable "network-subnet-cidr" {
+  type        = string
+  description = "The CIDR for the network subnet"
+  default     = "10.10.15.0/24"
+}
```
Comment (Dasomeone) [inline openstack/network-variables.tf:5]: newline

File: openstack/network-main.tf [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+  auto_create_subnetworks = "false"
+  routing_mode            = "GLOBAL"
+}
+# create public subnet
+resource "google_compute_subnetwork" "network_subnet" {
+  name          = "openstack-sample-app-subnet"
+  ip_cidr_range = var.network-subnet-cidr
+  network       = google_compute_network.vpc.name
+  region        = var.gcp_region
+}
```
Comment (Dasomeone) [inline openstack/network-main.tf:13]: newline

File: openstack/network-firewall.tf [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+resource "google_compute_firewall" "allow-ssh" {
+  name    = "openstack-sample-app-fw-allow-ssh"
+  network = google_compute_network.vpc.name
+  allow {
+    protocol = "tcp"
+    ports    = ["22"]
+  }
+  source_ranges = ["0.0.0.0/0"]
+  target_tags   = ["ssh"]
+}
```
Comment (Dasomeone) [inline openstack/network-firewall.tf:23]: newline

**Review by Dasomeone — APPROVED** [review]
LGTM, but we need a second approval from @rgeyer

File: openstack/bootstrap-vm.tp1 [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+sudo go build -buildvcs=false -o ./openstack-exporter .
+./openstack-exporter --os-client-config /etc/openstack/clouds.yaml devstack-admin > /dev/null 2>&1 &
+
+# Install Grafana agent
+sudo mkdir -p /etc/apt/keyrings/
+sudo wget -q -O - https://apt.grafana.com/gpg.key | gpg --dearmor | sudo tee /etc/apt/keyrings/grafana.gpg > /dev/null
+echo "deb [signed-by=/etc/apt/keyrings/grafana.gpg] https://apt.grafana.com stable main" | sudo tee /etc/apt/sources.list.d/grafana.list
+sudo apt-get update
+sudo apt-get install grafana-agent
+cat <<'CONFIG' | sudo tee /etc/grafana-agent.yaml
```
[the remark this answers is not in the document]
Comment (rgeyer) [reply openstack/bootstrap-vm.tp1:36]: Yes, and a lot of the "boilerplate" for cloud-init scripts, such as installing and configuring the agent should be shared snippets as well.

I think that this repo will eventually have a top-level library of these sorts of things which can be brought it as part of the automation.

**Comment by rgeyer** [comment]
> > @Dasomeone On the subject of Multipass, I was initially attempting to build this sample to run with Multipass but ran into issues with executing on an M2 Mac. Essentially the `nova` service that gets installed by devstack is somewhat supported, but not officially supported, making setup on a Multipass VM impossible for me to test on my machine.
> 
> That's a very fair point then! Might be worth adding a small known issues/platform support/etc section to the readme just to explain that, in case we attempt to convert it in the future

We may have to re-evaluate our use of multipass for this reason, or track the M2 Mac support more closely.

This was true for the "minikube" version of the multipass VM as well. Seems apple silicon support is taking it's sweet time in many ecosystems.

**Review by rgeyer — APPROVED** [review]
LGTM, sorry for gatekeeping this so long!

**Pull request merged**